pinot-controller/src/main/java/org/apache/pinot/controller/recommender/rules/impl/RealtimeProvisioningRule.java

```diff
@@ -0,0 +1,183 @@
+import org.apache.pinot.controller.recommender.io.InputManager;
+import org.apache.pinot.controller.recommender.realtime.provisioning.MemoryEstimator;
+import org.apache.pinot.controller.recommender.rules.AbstractRule;
+import org.apache.pinot.controller.recommender.rules.io.configs.IndexConfig;
+import org.apache.pinot.controller.recommender.rules.io.params.RealtimeProvisioningRuleParams;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.config.table.TableType;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.utils.DataSizeUtils;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+
+
+/**
+ * This rule recommends gives some recommendations useful for provisioning real time tables. Specifically it provides
+ * some recommendations for optimal segments size, total memory used per host, and consuming memory used per host based
+ * on the provided characteristics of the data.
+ */
+public class RealtimeProvisioningRule extends AbstractRule {
+  public static String OPTIMAL_SEGMENT_SIZE = "Optimal Segment Size";
+  public static String CONSUMING_MEMORY_PER_HOST = "Consuming Memory per Host";
+  public static String TOTAL_MEMORY_USED_PER_HOST = "Total Memory Used per Host";
+
+  private final RealtimeProvisioningRuleParams _params;
+
+  public RealtimeProvisioningRule(InputManager input, ConfigManager output) {
+    super(input, output);
+    _params = input.getRealtimeProvisioningRuleParams();
+  }
+
+  @Override
+  public void run()
+      throws InvalidInputException {
+
+    if (_params == null) {
+      // no realtime provisioning params provided; skip
+      return;
+    }
+
+    // prepare input to memory estimator
+    TableConfig tableConfig = createTableConfig(_output.getIndexConfig(), _input.getSchema());
+    long maxUsableHostMemoryByte = DataSizeUtils.toBytes(_params.getMaxUsableHostMemory());
+    int totalConsumingPartitions = _params.getNumPartitions() * _params.getNumReplicas();
+    int ingestionRate = (int) _input.getNumMessagesPerSecInKafkaTopic() / _params.getNumPartitions();
+    int[] numHosts = _params.getNumHosts();
+    int[] numHours = _params.getNumHours();
+
+    // run memory estimator
+    MemoryEstimator memoryEstimator =
+        new MemoryEstimator(tableConfig, _input.getSchema(), _input.getSchemaWithMetadata(),
+            (int) _input.getNumRecordsPerPush(), ingestionRate, maxUsableHostMemoryByte,
+            _params.getRealtimeTableRetentionHours());
+    File statsFile = memoryEstimator.initializeStatsHistory();
+    runAndRethrowIOException(() -> memoryEstimator
+        .estimateMemoryUsed(statsFile, numHosts, numHours, totalConsumingPartitions,
+            _params.getRealtimeTableRetentionHours()));
+
+    // extract recommendations
+    extractResults(memoryEstimator, numHosts, numHours, _output.getRealtimeProvisioningRecommendations());
+  }
+
+  private TableConfig createTableConfig(IndexConfig indexConfig, Schema schema) {
```

Comment:
I remember discussing this offline. 
Our internal integration with this tool will display all recommendations (indexing, partitioning, RT) once the user invokes it through the UI.
However, this work for recommending RT configs will first generate the segment right. To generate the segment, RT provisioning tool / MemoryEstimator will require to know the indexes. How are we taking care of this here ?

Reply:
We're using the output of other rules on the fly to create table config which is the input RT.Prov.Rule.